sigp/lighthouse-ui
# Add script to generate blsToExecutionChange for testnet validators.

Add script to generate blsToExecutionChange for testnet validators to make withdrawal testing easier on local testnet. The script does the following:
- Set up the parameters required for generating `blsToExecutionChange` using [staking-deposit-cli](https://github.com/ethereum/staking-deposit-cli)
- Runs `staking-deposit-cli` using docker and outputs the json to a local directory.
- Update local testnet to latest version to support Capella (copied from a Lighthouse branch)

To use this script, `lighthouse` and `lcli` needs to be built using this [Lighthouse PR](https://github.com/sigp/lighthouse/pull/3807), so that we could use a given mnemonic to set up the genesis validators. See steps below.

## Steps to test:

1. Check out the branch from [sigp/lighthouse#3808](https://github.com/sigp/lighthouse/pull/3807) and build lighthouse and lcli: `make && make install-lcli`
2. Pull latest changes on this PR
3. Start testnet  `./start_local_testnet.sh genesis.json` (make sure [`geth` & `bootnode`](https://geth.ethereum.org/downloads) are installed and available on `PATH`)
4. Generate a `blsToExecutionChange` json:
```
# generate for validator index 0 & 1
./generate_bls_to_execution_change.sh 0 0,1 <eth1_address>
```
5. Submit the `blsToExecutionChange` json to beacon node:
```
curl -X 'POST' \\n  'http://localhost:8001/eth/v1/beacon/pool/bls_to_execution_changes' \\n  -H 'accept: */*' \\n  -H 'Content-Type: application/json' \\n  --data "@bls_to_execution_changes/bls_to_execution_change-<your_file_name>.json"
```
6. Wait until Capella fork (about 1 minute) and the BLS change messages are processed. If you're interested, look for this message in the `beacon_node_<n>.log`, but it should happen pretty quickly if you submitted the message after Capella fork, so feel free to skip checking this:
```
INFO Processed BLS to execution change
```
7. Confirm `withdrawal_credentials` are updated to `0x01`:
```
curl -X GET "http://localhost:8001/eth/v1/beacon/states/head/validators/0" | jq .
{
  "execution_optimistic": false,
  "finalized": false,
  "data": {
    "index": "0",
    "balance": "32000001170",
    "status": "active_ongoing",
    "validator": {
      "pubkey": "0xa03c8f4fe1baf3b62ee1464778ae7a0bfe2fa3fe5baa454c4bfef35832b2378ecfe202a00cb05a34c2fc4c22bf6eff7c",
      "withdrawal_credentials": "0x01000000000000000000000049011adbcc3bc9c0307bb07f37dda1a1a9c69d2e",
      "effective_balance": "32000000000",
      "slashed": false,
      "activation_eligibility_epoch": "0",
      "activation_epoch": "0",
      "exit_epoch": "18446744073709551615",
      "withdrawable_epoch": "18446744073709551615"
    }
  }
}
```


NOTE: the local testnet is a copy from the Lighthouse repo, longer term we potentially want to unify them to reduce maintenance burden.
